pkg/pgmodel/ingestor/trace/trace_batcher.go

```diff
@@ -0,0 +1,248 @@
+	once            sync.Once
+	bufferedBatches chan Batch
+	wg              sync.WaitGroup
+	config          BatcherConfig
+}
+
+type BatcherConfig struct {
+	Batchers     int
+	Writers      int
+	MaxBatchSize int
+	BatchTimeout time.Duration
+}
+
+type WriteTraces func(context.Context, ptrace.Traces) error
+
+func NewBatcher(config BatcherConfig, writer Writer) *Batcher {
+	if config.Batchers == 0 {
+		config.Batchers = defaultBatchWorkers
+	}
+	if config.MaxBatchSize == 0 {
+		config.MaxBatchSize = DefaultBatchSize
+	}
+	if config.BatchTimeout == 0 {
+		config.BatchTimeout = DefaultBatchTimeout
+	}
+	if config.Batchers < 1 || config.Writers < 1 {
+		panic("number of batchers and writeres must be greater then zero")
+	}
+	bufferedBatches := make(chan Batch, defaultMaxBufferedBatches)
+	return &Batcher{
+		batchWriter:     newBatchWriter(config.Writers, writer, bufferedBatches),
+		in:              make(chan InsertTracesReq, defaultReqBufferSize),
+		stop:            make(chan struct{}, 1),
+		bufferedBatches: bufferedBatches,
+		config:          config,
+	}
+}
+
+func (b *Batcher) Run() {
+	for i := 0; i < b.config.Batchers; i++ {
+		b.wg.Add(1)
+		go func() {
+			defer b.wg.Done()
+			b.batch()
+		}()
+	}
+	b.batchWriter.run()
+}
+
+func (b *Batcher) batch() {
+	_, batcherSpan := tracer.Default().Start(context.Background(), "create-batches")
+	ticker := time.NewTicker(b.config.BatchTimeout)
+	batch := NewBatch(b.config.MaxBatchSize)
+	batcherSpan.AddEvent("New Batch")
+	flushBatch := func(batch *Batch) *Batch {
+		batchCp := *batch
+		b.bufferedBatches <- batchCp
+		batcherSpan.AddEvent("Batch sent to buffer")
+		batcherSpan.AddEvent("New Batch")
+		return NewBatch(b.config.MaxBatchSize)
+	}
+	for {
+		select {
+		case <-b.stop:
+			ticker.Stop()
+			if !batch.isEmpty() {
+				flushBatch(batch)
+			}
+			return
+		case item := <-b.in:
+			if batch.isFull() {
+				batch = flushBatch(batch)
+				ticker.Reset(b.config.BatchTimeout)
+			}
+			batch.add(item)
+		case <-ticker.C:
+			batcherSpan.AddEvent("Batch timeout reached")
+			if !batch.isEmpty() {
+				batch = flushBatch(batch)
+			}
+		}
+	}
+}
+
+func (b *Batcher) Stop() {
+	b.once.Do(func() {
+		b.stop <- struct{}{}
+		close(b.stop)
+		b.wg.Wait()
+		close(b.bufferedBatches)
+		b.batchWriter.stop()
+	})
+}
+
+// batchWriter writes batches using writer
+type batchWriter struct {
+	batches chan Batch
+	writers int
+	writer  Writer
+	stopCh  chan struct{}
+	once    sync.Once
+	wg      sync.WaitGroup
+}
+
+func newBatchWriter(writers int, writer Writer, batches chan Batch) *batchWriter {
+	return &batchWriter{
+		writer:  writer,
+		writers: writers,
+		batches: batches,
+		stopCh:  make(chan struct{}, 1),
+	}
+}
+
+func (bw *batchWriter) run() {
+	for i := 0; i < bw.writers; i++ {
+		bw.wg.Add(1)
+		go func() {
+			defer bw.wg.Done()
+			for {
+				select {
+				case b, ok := <-bw.batches:
+					if !ok {
+						return
+					}
+					bw.flush(b)
+				case <-bw.stopCh:
```

Comment:
I think we should handle these things via context as per convention?

Reply:
Not sure I follow. Isn't closing a channel a common way to stop goroutines?